### ISSUE TITLE
Introduce one-stop-shop endpoint for updates

### ISSUE THREAD

**Issue by tushev** [AdoptOpenJDK/openjdk-api-v3#367]
[As suggested](https://github.com/AdoptOpenJDK/openjdk-api-v3/issues/206#issuecomment-725662689) by @aahlenst in #206, please add a 'one-stop-shop' endpoint for updating JDK/JRE instances.

Suggested name: `/v3/updates/query`

The suggested endpoint should receive a list of versions installed on a user's machine:
```
{
    "versions": [
         {
            "version_string": "11.0.9.1+11.1".
            "os": "windows",
            "image_type": "jre",
            "arch": "x64",
            "jvm_implementation": "openj9",
            "heap_size": "large",
            "vendor": "adopt" // not sure if this is required
        },
        {
            "version_string": "11.0.9.101",
            "os": "windows",
            "image_type": "jdk",
            "arch": "x32",
            "jvm_implementation": "hotspot"
        }
   ],
   // some other params if needed
}
```

In response, the endpoint should provide a corresponding JSON array of versions, with the following properties/fields:
* **boolean** `eol_reached`, mandatory
* **boolean** `is_lts`, mandatory
* **boolean** `update_available`, mandatory
* **[Release object](https://github.com/AdoptOpenJDK/openjdk-api-v3/blob/master/adoptopenjdk-api-v3-models/src/main/kotlin/net/adoptopenjdk/api/v3/models/Release.kt)** `latest_release`, optional
Present when update is available, provides all the data for performing the update(s) without extra API requests<br>(JSON object like in `/v3/assets/latest/{feature_version}/{jvm_impl}`, please correct me if this is not `Release object`)

<hr>

Another question is: in which format should the versions be provided. Currently locally installed instances use the following formats :
* 'registry/dir format', like `11.0.9.101`. Very much like semver, available in Windows Registry.
* version string, like `11.0.9.1+11.1`. Available in `release` file

In many cases only one of this version formats is available to the updater, at least until [this issue](https://github.com/AdoptOpenJDK/openjdk-build/issues/2248#issuecomment-732980755) is resolved. 

IMO, the suggested endpoint should be able to parse any version format, as well as this one:
* 'feature-only', aka privacy-enhanced mode: `11`
In this case the endpoint should proceed as usual, returning the latest release for each version requested.

**Comment by johnoliver** [comment]
Would the request:

https://api.adoptopenjdk.net/v3/assets/version/%5B11%2C12%29?architecture=x64&heap_size=normal&image_type=jdk&jvm_impl=hotspot&os=windows&page=0&page_size=1&project=jdk&release_type=ga&sort_method=DEFAULT&sort_order=DESC&vendor=adoptopenjdk

Satisfy your requirements?

This is calling the `/assets/version` endpoint with the version range `[11,12)` (i.e `11 <= version < 12`), and the parameters for the binary you want, and selecting only the most recent.

Alternatively we could add more query parameters to the `/assets/latest` endpoint that may also address your use case.

If you can give your use case we can help to see if there is an alternative, but I believe that what you are trying to achieve should be addressed by multiple calls to the `/assets/version` as described above. 

If you wish to try to reduce the number of web requests made I would not worry about that at all, it will probably be as much work for both the client to form the request, and the server to parse and validate the request than just making multiple requests. Do you think you have a use case where multiple requests would be an issue?

**Comment by tushev** [comment]
The key ideas are:

* Combine multiple requests to `/assets/latest` into one
* If the user opted in to transmit full version strings (like `11.0.9.1+1`), perform version comparison on the server side, so [issues like described in this comment](https://github.com/AdoptOpenJDK/openjdk-installer/issues/4#issuecomment-727928039) could be addressed without having to update the updater (and wait until its new version propagates to end users)
* When #206 will be implemented, eliminate the necessity for extra request to `info/available_releases` to obtain `eol_reached` data for all the releases

> Do you think you have a use case where multiple requests would be an issue?

From within my network, the query to `/assets/latest` looks like (in browser):
![image](https://user-images.githubusercontent.com/18406797/110273415-57634680-7fee-11eb-9518-dcbdd289cd02.png)
I did not measured the timings from within the C# app, but it does not feel any faster :). Generally, with 1-2 Adopt installations, it takes a couple of seconds to check for updates. However, when there are multiple installs, it may take tens of seconds to check all of them for updates. With #206, it would mean +1 request.

![image](https://user-images.githubusercontent.com/18406797/110272023-fd14b680-7fea-11eb-9030-58afba703ef5.png)
This is definitely a synthetic/devenv case, but I'm aware that there are users with multiple Adopt installations. However, I cannot say anything about percentage - my updater does not collect any data.

It is absolutely not a problem for me to perform multiple API requests (that's what I'm already doing, the API has (or will have soon) all the necessary functions). The key idea is to reduce the load on API servers and eliminate extra connection/TLS handshakes. However, it's up to you to decide whether it's worth the work required to implement this...

> Would the request: < ... assets/version ...> satisfy your requirements?
> This is calling the `/assets/version` endpoint with the version range `[11,12)` (i.e `11 <= version < 12`), and the parameters for the binary you want, and selecting only the most recent.
> Alternatively we could add more query parameters to the `/assets/latest` endpoint that may also address your use case.

Currently I'm using `/assets/latest` and filtering the binaries on the client side. The only difference is `vendor` field, however, @M-Davies is working on introducing it to `/assets/latest` in #365

**Comment by johnoliver** [comment]
From the servers point of view I would not worry about the load, it would be preferable to have the multiple connections over batching up requests. 

In terms of speeding things up, its not a great answer as it would probably increase the complexity on your end but you can make those requests in parallel and that is not a problem for the API. We should probably track down exactly where the request time is coming from and if there is any scope for reducing it.

**Comment by karianna** [comment]
@tushev Unfortunately we won't be adding this endpoint to the API.  We'll raise as separate issue for the performance of the end points
